### DISCUSSION TITLE
How do I collect ALL output from pexpect.spawn()

### DISCUSSION THREAD

**Discussion by KeithProctor** [pexpect/pexpect#775] [General]
Currently the c.read() function only returns values from output of the program that I am calling and does not collect what happens during the interactive portion.  (IOW:  Waiting for user name and password).  Since errors can happen during that section I need to collect all output after the cmd is called.

Using the run command works fine.  I am trying to convert to using expect.spawn() since that looks very close to the ssh version of expect.  Current the run() method works fine in that it works locally.  I can't find getOutput() type function that would return the entire output.

    import pexpect
    class Expect( ):
        def Do( self, cmd, program: list = [], timeout: int = 20 ):
            # result = run( cmd, events=program, timeout=timeout).decode() 

             result = ''
             c = pexpect.spawn( cmd, encoding='utf-8', timeout=timeout )
             for curItem in program:
                # print( 'wait: ' + curItem[0] + ' resp: ' + curItem[1])
                 c.expect( curItem[0] )
                 c.sendline( curItem[1] )
                # result += c.read()
		
              result += c.read()

              print ( 'res: ' + str(result) )
              return result

     cmd  = 'fmsadmin list files -s'
     prog = [('username \\(.+\\):', 'yourUN\n'), ('password:', 'yourPW\n')]
     res  = Expect().Do( cmd, prog )    # Returns everything
    # use results to verify functionality

**Comment by KeithProctor** [comment]
@jquast @bukepo @edcolvin @EmersonPrado - Very sorry to connect with you guys directly.  I've posted this same question on StackOverflow and didn't get any bites so I came here and posted the same question.  I'm hoping that you guys can help me resolve this question?

**Comment by EmersonPrado** [comment]
@KeithProctor Could you pls post the link to the StackOverflow question? This doesn't seem like a Pexpect issue, so I guess we better talk there.
Regards,
Emerson

**Comment by KeithProctor** [comment]
@EmersonPrado
Hey... TY so much for helping.  Definitely a Pexpect issue.  More people will learn on Stack overflow so here is that link:

https://stackoverflow.com/questions/67238593/how-do-i-collect-all-output-from-pexpect-spawn

**Comment by KeithProctor** [comment]
@EmersonPrado Thanks for offering to help.  I was wondering if we could have a live meeting.  Possibly through meeting maker?

**Comment by jquast** [comment]
This question was self-answered on SO where it was duplicated: **use `c.before`** and maybe `c.after`, this provides all output that occurred during the "interactive portion", the result of sendline() and output leading up to a match with expect(). https://stackoverflow.com/questions/67238593/how-do-i-collect-all-output-from-pexpect-spawn/67306955#67306955

That leaves only that the password is not "captured" as output, because it is not in output, the terminal has changed modes to ensure that input is not echoed back to output, for example in python std lib, https://github.com/python/cpython/blob/42b6883d5f648a95e33f815364c8a529e4574caa/Lib/getpass.py#L71

Because your code sends the password, you may just conditionally capture a copy of your input into the result when `curItem[0] == password`, or a more versatile case by inspecting the terminal flags of the child pty.
